lib/rucio/core/account_counter.py

```diff
@@ -82,7 +85,7 @@ def del_counter(rse_id, account, *, session: "Session"):
 
 
 @read_session
-def get_updated_account_counters(total_workers, worker_number, *, session: "Session"):
+def get_updated_account_counters(total_workers: int, worker_number: int, *, session: "Session") -> list["Query"]:
```

Comment:
This was the reference I used for that: https://stackoverflow.com/questions/55361186/type-hinting-sqlalchemy-query-result

> [Query.all()](https://github.com/sqlalchemy/sqlalchemy/blob/201c4a60e4b8af56d9c02a3675d1443ba4171c89/lib/sqlalchemy/orm/query.py#L3155-L3161) calls list() on the Query object itself

I'm not 100% sure about it either, but I'm not sure about what other type this would be.

Reply:
But looking at the query:

https://github.com/rucio/rucio/blob/cc0067cd2e5ae1c2c6c69cd8adc401db1bdcf015/lib/rucio/core/account_counter.py#L94

I would have guessed something like `list[tuple[InternalAccount, str]]`. Did you try it and saw errors?

Reply:
I have not tried `list[tuple[InternalAccount, str]]` yet - I'll try it now to see if it works

Reply:
We need to have a closer look at this. `.all()` returns instances of `sqlalchemy.engine.row.Row`. I’m thinking it would be preferable to:

 1. Convert the query to SQLAlchemy 2.0 syntax (unrelated to hints, but a good opportunity)
 2. Return dictionaries with `return [row._asdict() for row in query.all()]`
 3. Adapt `lib/rucio/daemons/abacus/account.py`

My motivation for this proposal is that the use of indices feels poor:

https://github.com/rucio/rucio/blob/cc0067cd2e5ae1c2c6c69cd8adc401db1bdcf015/lib/rucio/daemons/abacus/account.py#L72-L73

What do you think?

And on a side note, do you understand the usefulness of this? I think it’s identical and can go in `filter_thread_work()` directly.

https://github.com/rucio/rucio/blob/cc0067cd2e5ae1c2c6c69cd8adc401db1bdcf015/lib/rucio/core/account_counter.py#L97-L100

It was introduced in f3a073aa9da181070d6b41572838d28727528dd6.

Reply:
I've done the first three changes suggested in the latest revision, but I think they're outside the scope of this PR, and I think the focus should be kept on typing (and make separate PRs for these behavioural changes when needed, which I think might happen fairly frequently with these typing-focused PRs) 

On `filter_thread_work`: I agree that it could be refactored like that, but it seems like that method is used in quite a few other places and I don't think refactoring would fit the scope of this PR - best to open an issue for it and we can work on it separately.

Reply:
Yes and no.

The SQLAlchemy 2.0 syntax, I admit, was over the top. There are five trivial queries in that file; they can all be worked on at the same time in a PR which aims to do it from the start. In retrospect, I should not have asked you to do it. I’ll let you choose whether to keep it or not. Note, however, that you’ll have to follow the coding style we’ve come to adopt for queries:

```python
query = select(
    models.UpdatedAccountCounter.account,
    models.UpdatedAccountCounter.rse_id,
).distinct(
)
```

Now, the change in the return type of `get_updated_account_counters()` *is* very much related to this PR. I’ve asked you to do it because I believe `list[tuple[InternalAccount, str]]` is incorrect. So we had three options:

 1. Not add type hints.
 2. Add incorrect type hints.
 3. Adjust and add correct type hints.

I chose option (3). However, you cannot hide it under `Auth: Migrate to SQLAlchemy 2.0 syntax; #6057`. You have to split it into its own commit.

Now, I did not understand your comment about `filter_thread_work()`. That function is completely fine as it is. What I am asking you to do is to:

 1. Remove those four lines.
 2. Replace `hash_variable=hash_variable` with `hash_variable='CONCAT(account, rse_id)'`.

In a separate commit, of course. Yes, it’s not related to type hints, but unlike the SQLAlchemy 2.0 syntax migration, there will be no dedicated issue or pull request to fix it (because it doesn’t happen anywhere else).

Reply:
I had misunderstood regarding `filter_thread_work()`, I thought you meant that this logic:

https://github.com/rucio/rucio/blob/cc0067cd2e5ae1c2c6c69cd8adc401db1bdcf015/lib/rucio/core/account_counter.py#L97-L100

could have gone directly within the `filter_thread_work()` function, rather than being done outside of it; now I understood that you meant that the `CONCAT` command is case insensitive and thus the `if/else` block is not needed. I'll make the change you proposed now